src/project/internal/recentprojectscontroller.cpp

```diff
@@ -0,0 +1,214 @@
+        if (!m_reloadingBlocked) {
+            m_dirty = true;
+
+            {
+                std::lock_guard lock(m_thumbnailCacheMutex);
+                m_thumbnailCache.clear();
+            }
+        }
+
+        m_recentListChanged.notify();
+    });
+}
+
+const RecentFilesList& RecentProjectsController::recentProjectsList() const
+{
+    TRACEFUNC;
+
+    if (m_dirty) {
+        const_cast<RecentProjectsController*>(this)->loadRecentProjectsList();
+    }
+
+    const_cast<RecentProjectsController*>(this)->removeNonexistentFiles();
+
+    return m_recentProjectsList;
+}
+
+Notification RecentProjectsController::recentProjectsListChanged() const
+{
+    return m_recentListChanged;
+}
+
+void RecentProjectsController::prependRecentFile(const RecentFile& newFile)
+{
+    if (newFile.empty()) {
+        return;
+    }
+
+    TRACEFUNC;
+
+    RecentFilesList newList;
+    newList.reserve(m_recentProjectsList.size() + 1);
+    newList.push_back(newFile);
+
+    for (const RecentFile& file : m_recentProjectsList) {
+        if (file != newFile && fileSystem()->exists(file)) {
+            newList.push_back(file);
+        }
+    }
+
+    setRecentProjectsList(newList, true);
+
+    prependPlatformRecentFile(newFile);
+}
+
+void RecentProjectsController::clearRecentFiles()
+{
+    setRecentProjectsList({}, true);
+
+    clearPlatformRecentFiles();
+}
+
+void RecentProjectsController::prependPlatformRecentFile(const io::path_t&) {}
+
+void RecentProjectsController::clearPlatformRecentFiles() {}
+
+void RecentProjectsController::loadRecentProjectsList()
+{
+    ByteArray data = configuration()->rawRecentProjectsData();
+
+    RecentFilesList newList;
+
+    DEFER {
+        setRecentProjectsList(newList, false);
+    };
+
+    if (data.empty()) {
+        return;
+    }
+
+    QJsonParseError err;
+    QJsonDocument jsonDoc = QJsonDocument::fromJson(data.toQByteArrayNoCopy(), &err);
+    if (err.error != QJsonParseError::NoError) {
+        return;
+    }
+
+    if (!jsonDoc.isArray()) {
+        return;
+    }
+
+    for (const QJsonValue val : jsonDoc.array()) {
+        newList.push_back(val.toString());
+    }
+}
+
+void RecentProjectsController::removeNonexistentFiles()
+{
+    bool removed = false;
+
+    RecentFilesList newList;
+    newList.reserve(m_recentProjectsList.size());
+
+    for (const RecentFile& file : m_recentProjectsList) {
+        if (fileSystem()->exists(file)) {
+            newList.push_back(file);
+        } else {
+            removed = true;
+        }
+    }
+
+    if (removed) {
+        setRecentProjectsList(newList, false);
+
+        async::Async::call(nullptr, [this, newList]() {
+            saveRecentProjectsList();
+        });
+    }
+}
+
+void RecentProjectsController::setRecentProjectsList(const RecentFilesList& list, bool save)
+{
+    m_recentProjectsList = list;
+
+    {
+        std::lock_guard lock(m_thumbnailCacheMutex);
+        m_thumbnailCache.clear();
```

Comment:
Could you please explain why we need this? So the cache is cleared every time when the user opens a score. Is my assumption correct?

Reply:
This is indeed what would happen. This matches the old behaviour of regenerating all thumbnails every time when something changes in the list of recent files. This ensures that the user will see up-to-date thumbnails every time after saving a file in some instance. However, we could make this more sophisticated, for example by not invalidating the whole cache, but checking the "last modified" date of the file in question. I'll investigate.